### PR TITLE
4. Refactor Publish

### PR DESCRIPTION
**Pull request by timoxley**
Co-locating and re-jigging publishing step.

Also:

* Fixes non-functional MessageCreationUtil tests.
* Swaps out receptacle for more general purpose & reusable `mem`, `p-memoize` & `quick-lru`
* Removes expensive `ethers.Wallet.createRandom()` call in tests.
* Tidied up login endpoint tests.
* Removed all use of `ensureConnected`/`ensureDisconnected`.

Builds on:
* Split Client into Publisher/Subscriber/Resender https://github.com/streamr-dev/streamr-client-javascript/pull/156
* Connection/Reconnection Refactoring https://github.com/streamr-dev/streamr-client-javascript/pull/160
* Publish/Subscribe Promisification https://github.com/streamr-dev/streamr-client-javascript/pull/162

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -35,3 +38,35 @@ export function waitFor(emitter, event) {
         emitter.once('error', onError)
     })
 }
+
+/* eslint-disable object-curly-newline */
+export function CacheAsyncFn(fn, {
```
Comment (mirpo) [inline src/utils.js:43]: 👍

**Review by mirpo — APPROVED** [review]
lgtm

File: src/Publisher.js [hunk 2]
```diff
@@ -18,12 +23,157 @@ function getStreamId(streamObjectOrId) {
...
+    clear() {
+        this.messageRefs.clear()
+    }
+}
+
+export class MessageCreationUtil {
+    constructor(client) {
+        this.client = client
+        const cacheOptions = client.options.cache
+        this.msgChainer = new OrderedMessageChainCreator(cacheOptions)
```
Comment (harbu) [inline src/Publisher.js:63]: Constructor of `OrderedMessageChainCreator` takes no arguments?

**Review by harbu — APPROVED** [review]
LGTM 👍

**Pull request merged**